runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/StreamingRequestBody.kt

```diff
@@ -20,44 +21,50 @@ import kotlin.coroutines.CoroutineContext
 /**
  * OkHttp [RequestBody] that reads from [body] channel or source
  */
+@OptIn(DelicateCoroutinesApi::class, ExperimentalStdlibApi::class)
 internal class StreamingRequestBody(
     private val body: HttpBody,
-    callContext: CoroutineContext,
-) : RequestBody(), CoroutineScope {
+    private val callContext: CoroutineContext,
+) : RequestBody() {
 
     init {
         require(body is HttpBody.ChannelContent || body is HttpBody.SourceContent) { "Invalid streaming body $body" }
     }
 
-    private val producerJob = Job(callContext[Job])
-    override val coroutineContext: CoroutineContext = callContext + producerJob + callContext.derivedName("send-request-body") + Dispatchers.IO
     override fun contentType(): MediaType? = null
     override fun contentLength(): Long = body.contentLength ?: -1
     override fun isOneShot(): Boolean = body.isOneShot
     override fun isDuplex(): Boolean = body.isDuplex
 
-    override fun writeTo(sink: BufferedSink) = try {
-        doWriteTo(sink)
-    } catch (ex: Exception) {
-        throw when (ex) {
-            is IOException -> ex
-            // wrap all exceptions thrown from inside `okhttp3.RequestBody#writeTo(..)` as an IOException
-            // see https://github.com/awslabs/aws-sdk-kotlin/issues/733
-            else -> IOException(ex)
+    override fun writeTo(sink: BufferedSink) {
+        try {
+            doWriteTo(sink)
+        } catch (ex: Exception) {
+            when (ex) {
+                is CancellationException -> {
+                    callContext.trace<StreamingRequestBody> { "request cancelled" }
+                    // shouldn't need to propagate the exception because okhttp is cancellation aware via executeAsync()
+                    return
+                }
+                is IOException -> throw ex
+                // wrap all exceptions thrown from inside `okhttp3.RequestBody#writeTo(..)` as an IOException
+                // see https://github.com/awslabs/aws-sdk-kotlin/issues/733
+                else -> throw IOException(ex)
+            }
         }
     }
 
-    @OptIn(ExperimentalStdlibApi::class)
     private fun doWriteTo(sink: BufferedSink) {
+        val context = callContext + callContext.derivedName("send-request-body")
         if (isDuplex()) {
             // launch coroutine that writes to sink in the background
-            launch {
+            GlobalScope.launch(context + Dispatchers.IO) {
                 sink.use { transferBody(it) }
             }
```

Comment:
Without passing existing context/job yes but we are passing the `callContext` to it so the coroutine will be launched as a child. IIRC cancellation is covered in the tests.

Reply:
Yes but now that the code no longer throws an exception, the cancellation test doesn't assert anything. As long as nothing crashes the test passes.

Reply:
I think this is still fine. The test will fail if the implementation fails to respect the cancellation which is the purpose of the test. The assertion on a specific exception was only there to ensure we wrapped it correctly, now that we are swallowing one there isn't anything to assert there.

This is the meat of the test really:

```kotlin
        job.cancel()
        withTimeout(2.seconds) { // See https://github.com/awslabs/smithy-kotlin/issues/739
            // writeTo() should end up blocked waiting for data that will never come.
            // If the job used in the implementation isn't tied to the parent coroutine correctly
            // it will block forever
            job.join()
        }
```